libs/checkpoint-sqlite/src/index.ts

```diff
@@ -165,19 +165,80 @@ CREATE TABLE IF NOT EXISTS writes (
     config: RunnableConfig,
     options?: CheckpointListOptions
   ): AsyncGenerator<CheckpointTuple> {
-    const { limit, before } = options ?? {};
+    const { limit, before, filter } = options ?? {};
     this.setup();
     const thread_id = config.configurable?.thread_id;
-    let sql = `SELECT thread_id, checkpoint_ns, checkpoint_id, parent_checkpoint_id, type, checkpoint, metadata FROM checkpoints WHERE thread_id = ? ${
-      before ? "AND checkpoint_id < ?" : ""
-    } ORDER BY checkpoint_id DESC`;
-    if (limit) {
-      sql += ` LIMIT ${limit}`;
+    const checkpoint_ns = config.configurable?.checkpoint_ns;
+
+    const withClause =
+      filter && Object.keys(filter).length
+        ? "WITH augmented_checkpoints AS (\n" +
+          "  SELECT \n" +
+          "    ck.thread_id,\n" +
+          "    ck.checkpoint_ns,\n" +
+          "    ck.checkpoint_id,\n" +
+          "    jsonb(CAST(metadata as TEXT)) AS json_metadata\n" +
```

Comment:
Why do we need this separately vs. just returning it as a field in the initial query

Reply:
IIRC I didn't want to modify the structure of the result set returned by the query (`CheckpointRow[]`), as ultimately I was only doing this to support a filtering operation. That said, I can't recall why I opted for a CTE here when  I could've likely just done the following:

```
    whereClause.push(
      ...Object.entries(filter ?? {})
        .filter(([_, value]) => value !== undefined)
        .map(([key]) => `jsonb(CAST(metadata as TEXT))->>'$.${key}' = ?`)
    );
```

Of course, the ideal chang would be to migrate `checkpoint.metadata` to be a `jsonb` column rather than a blob. If it's all the same to you, my preference would be to avoid the migration now, but add a note to #592 to change up the column type at the same time, as #592 necessitates a migration anyway.

Reply:
Also I I just realised that I'm appending the filter object's `key` field directly into the query. This could become a SQL injection vulnerability (e.g. in an implementation that defines the filter in frontend code and passes it through an API unchecked). I'll add an `Array.filter` step into this to check the key names so that can't happen.

Reply:
Yeah suppose that could happen - good thinking